### ISSUE TITLE
Scaffolding: generate or hydrate?

### ISSUE THREAD

**Issue by sitch** [ash-project/ash#137]
I was creating a sample project to play around with and noticed there wasn't any kind of scaffold generators.

I'd be happy to open a PR for this but I wanted to know your thoughts @zachdaniel 

## 1. Add Generator Task

add a `Mix.Task` that can be called by `mix ash.new ...` that generates an ash project a la `mix phx.new`

### Pros
- Easy enough to copy paste from Phx.New
- Can easily allow generators to pick up user templates in `priv/templates`

### Cons
- Hard to maintain sane/readable templates with eex (https://www.hygen.io/ is probably the best i've seen, but even still these things are rigid and messy)

## 2. Add Snapshot (De)hydration Tasks

add a `Mix.Task` that can be called by `mix ash.hydrate snapshot_file`.  

This can either write actual elixir resource files or generate/compile/track beam files via ast inference

add a `Mix.Task` that can be called by `mix ash.dehydrate --out snapshot_file`.  

This can rm resource files in favor of hash tracked beam files

### Pros
- Can leverage some snapshot code I saw at least in the psql adapter
- Can use versioned & timestamped snapshots (and maybe help one person debug with `git bisect`)
- This type of dynamic beam file writing can actually be very conducive to solving notoriously slow macro based compiles (i.e. absinthe) by generating independent/self contained (sub) applications

### Cons
- Requires the snapshot file layout to be long term stable
- Goes _wayy_ deeper into WTF/magic/metaprogramming code territory

## Final thoughts

Let me know if the hydration (if thats even a good term?) stuff makes sense -- the main idea is that a categorical type hypergraph is superior to _actual_ elixir files (provided they are largely boilerplate, macro-constructed and serializable)..

This can afford simple mix tasks to expand/dilate & write files that need manual editing, as well as mix tasks to remove any elixir source files that are deemed pure boilerplate, in favor of some sort of serializable snapshot format.

This makes playing with things like json schema, https://graphql-code-generator.com, etc super easy to interop with as the fundamental description is serializable.

Lastly, this mostly makes irrelevant the need for macro level inference via #136 as those actual elixir files could just be substituted with either the compiled beam file or a use macro which reconstitutes the boilerplate ast from the snapshot

**Comment by zachdaniel** [comment]
Hey @sitch another really great idea. The idea of snapshots has been rolling around in my head for a while, specifically in relation to versioning resources. How neat would it be if we had a `to_snapshot/1` or similar for each DSL entity/section that allowed it to be dumped to json? We could then additionally support a `from_snapshot/1`. The big benefit here in my mind is that an Ash resource definition could (eventually, would have to stabilize first) be a cross language concept if someone wanted to write an engine in a different language. I think there is a lot of power in being able to "hydrate" these snapshots, as it could actually allow for a sort of "time travel" if we make the data layer smart enough. E.g your clients could send a version header, and always get that version of resources. At some point you'd need to retire versions for sure, but I think we could provide that at a pretty low implementation cost. 

However, I think that "hydrating" a resource from a snapshot and generating resources from the CLI are relatively separate concerns. The main reason I haven't written generators is because you'd have to type roughly the same amount of characters if you write it into the command line as you would if you just wrote the resource. Not that I don't find value in generators, I think it would be a good tool solely to get people up and running w/ a skeleton resource, the juice just hasn't been worth the squeeze for me yet (so much to do).

Ultimately the idea of making these snapshots actually the source of truth is very interesting if it can be abstracted away from the user in a painless way. I will say this much: Having to dance around the elixir compiler has been a huge pain. Its the main reason that the step is also broken into a DSL and a "Transformer" phase. But being an Elixir DSL we get a lot "for free", e.g its all self-documented on hex docs (because I'm generating modules and functions). 

As I'm writing all of this, it strikes me that there is probably not one right answer, and that ultimately the introspection callbacks don't really need to care (on purpose). We load things into `:persistent_term` instead of using module attributes to prepare for exactly this kind of eventuality. There will need to be some places where we assume that the resource is a module name that have to be fixed. But other than that, saying `Ash.Resource.attributes(resource)` really just boils down to a call to `:persistent_term.get(resource, ...)`. So most of the system already doesn't care how that information got there. The biggest challenge will be that there are module/function references throughout resources, for things like custom policies/custom validations. That part wouldn't really ever be "portable".

I've got a lot of "boots on the ground" work left to do so I can't really engage with some of this fun stuff yet, but I welcome you to take a stab at supporting any of the things we've discussed here.

Any of these things sound promising/useful on their own:

* taking resource snapshots (e.g json)
* mix tasks to generate DSL | mix tasks to generate snapshots
* a task to "hydrate" a resource snapshot -> DSL
* resource snapshots being used in the raw as resources
